src/transforms/window.js

```diff
@@ -66,48 +64,38 @@ function maybeReduce(reduce = "mean") {
   return reduceSubarray(reduce);
 }
 
-function looseReducer(reducer) {
-  return (k, s) => {
-    const reduce = reducer(k, s);
-    return {
-      map(I, S, T) {
-        const n = I.length;
-        reduce.map(I, S, T, true);
-        for (let i = 0; i < s; ++i) {
-          const j = Math.min(n, i + k - s);
-          reducer(j, i).map(slice(I, 0, j), S, T);
-        }
-        for (let i = n - k + s + 1; i < n; ++i) {
-          const j = Math.max(0, i - s);
-          reducer(n - j, i - j).map(slice(I, j, n), S, T);
-        }
-      }
-    };
-  };
-}
-
-function slice(I, i, j) {
-  return I.subarray ? I.subarray(i, j) : I.slice(i, j);
-}
-
+// Note that the subarray may contain NaN even in the strict case; we expect the
+// function f to handle that itself (e.g., by filtering as needed). The D3
+// reducers (e.g., min, max, mean, median) already handle NaN input, so it’s
+// faster to avoid redundant filtering.
 function reduceSubarray(f) {
-  return (k, s) => ({
-    map(I, S, T, skipNaN) {
+  return (k, s, strict) => strict ? ({
+    map(I, S, T) {
       const C = Float64Array.from(I, i => S[i] === null ? NaN : S[i]);
       let nans = 0;
       for (let i = 0; i < k - 1; ++i) if (isNaN(C[i])) ++nans;
       for (let i = 0, n = I.length - k + 1; i < n; ++i) {
         if (isNaN(C[i + k - 1])) ++nans;
-        T[I[i + s]] = nans === 0 ? f(C.subarray(i, i + k)) : skipNaN ? f(C.subarray(i, i + k).filter(d => !isNaN(d))) : NaN;
+        T[I[i + s]] = nans === 0 ? f(C.subarray(i, i + k)) : NaN;
         if (isNaN(C[i])) --nans;
       }
     }
+  }) : ({
+    map(I, S, T) {
+      const C = Float64Array.from(I, i => S[i] === null ? NaN : S[i]);
+      for (let i = -s; i < 0; ++i) {
+        T[I[i + s]] = f(C.subarray(0, Math.max(0, i + k)));
```

Comment:
If we assume that s <= k (which should be true), then we can remove the Math.max here, right?

```suggestion
        T[I[i + s]] = f(C.subarray(0, i + k));
```